src/python_testing/TC_WASHERCTRL_2_1.py

```diff
@@ -0,0 +1,130 @@
+
+    def steps_TC_WASHERCTRL_2_1(self) -> list[TestStep]:
+        steps = [
+            TestStep(1, "Commissioning, already done",
+                     is_commissioning=True),
+            TestStep(2, description="TH reads from the DUT the SpinSpeeds attribute",
+                     expectation="Verify that the DUT response contains a list of strings. The maximum size of the list is 16."),
+            TestStep(3, description="TH reads from the DUT the SpinSpeedCurrent attribute",
+                     expectation="Verify that the DUT response contains a uint8 with value between 0 and numSpinSpeeds-1 inclusive."),
+            TestStep(4, description="TH writes a supported SpinSpeedCurrent attribute that is a valid index into the list"
+                     + "of spin speeds (0 to numSpinSpeeds - 1) and then read the SpinSpeedCurrent value",
+                     expectation="Verify DUT responds w/ status SUCCESS(0x00) and the SpinSpeedCurrent value was set accordingly"),
+            TestStep(5, description="TH writes an unsupported SpinSpeedCurrent attribute that is other than 0 to DUT",
+                     expectation="Verify that the DUT response contains Status CONSTRAINT_ERROR response")
+        ]
+
+        return steps
+
+    @run_if_endpoint_matches(has_feature(Clusters.LaundryWasherControls,
+                                         Clusters.LaundryWasherControls.Bitmaps.Feature.kSpin))
+    async def test_TC_WASHERCTRL_2_1(self):
+
+        endpoint = self.get_endpoint(default=1)
+
+        self.step(1)
+
+        # Read the SpinSpeeds attributes
+        self.step(2)
+        list_speed_speeds = await self.read_single_attribute_check_success(endpoint=endpoint,
+                                                                           cluster=Clusters.Objects.LaundryWasherControls,
+                                                                           attribute=Clusters.LaundryWasherControls.Attributes.SpinSpeeds)
+
+        asserts.assert_true(isinstance(list_speed_speeds, list), "Returned value was not a list")
+        numSpinSpeeds = len(list_speed_speeds)
+        asserts.assert_less_equal(numSpinSpeeds, MAX_SPIN_SPEEDS, "List of SpinSpeeds larger than maximum allowed")
+
+        # Read the SpinSpeedCurrent attribute
+        self.step(3)
+        spin_speed_current = await self.read_single_attribute_check_success(endpoint=endpoint,
+                                                                            cluster=Clusters.Objects.LaundryWasherControls,
+                                                                            attribute=Clusters.LaundryWasherControls.Attributes.SpinSpeedCurrent)
+        asserts.assert_true(isinstance(spin_speed_current, int), "SpinSpeedCurrent has an invalid value")
+        asserts.assert_true(0 <= spin_speed_current <= (numSpinSpeeds - 1), "SpinSpeedCurrent outside valid range")
+
+        self.step(4)
+        for requested_speed in range(0, numSpinSpeeds - 1):
```

Comment:
Looks like off by one to me as range should already apply a -1

```suggestion
        for requested_speed in range(0, numSpinSpeeds):
```